src/libpiano/response.c

```diff
@@ -241,6 +316,173 @@ PianoReturn_t PianoResponse (PianoHandle_t *ph, PianoRequest_t *req) {
+
+			if (!json_object_object_get_ex (result, "items", &playlists)) {
+				break;
+			}
+
+			for (int i = 0; i < json_object_array_length (playlists); i++) {
+				PianoStation_t *tmpStation;
+				json_object *s = json_object_array_get_idx (playlists, i);
+
+				if ((tmpStation = calloc (1, sizeof (*tmpStation))) == NULL) {
+					return PIANO_RET_OUT_OF_MEMORY;
+				}
+				tmpStation->stationType = PIANO_TYPE_PLAYLIST;
+
+				PianoJsonParsePlaylist(s, tmpStation);
+
+				/* start new linked list or append */
+				ph->stations = PianoListAppendP (ph->stations, tmpStation);
+			}
+			break;
+		}
+
+	// Get album or playlist tracks
+		case PIANO_REQUEST_GET_TRACKS: {
+			assert (req->responseData != NULL);
+
+			PianoRequestDataGetPlaylist_t *reqData = req->data;
+			PianoSong_t *playlist = NULL;
+			PianoSong_t *FullPlaylist = NULL;
+
+			assert (result != NULL);
+			assert (req->responseData != NULL);
+			assert (reqData != NULL);
+
+			switch(reqData->station->stationType) {
+				case PIANO_TYPE_PLAYLIST: {
+					json_object *tracks = NULL;
+					json_object *annotations = NULL;
+					if (!json_object_object_get_ex (result, "tracks", &tracks)) {
+						break;
+					}
+					assert (tracks!= NULL);
+					LOG("got tracks\n");
+					if (!json_object_object_get_ex (result, "annotations", &annotations)) {
+						break;
+					}
+					assert (annotations != NULL);
+					LOG("got annotations\n");
+
+					for (int i = 0; i < json_object_array_length (tracks); i++) {
+						json_object *s = json_object_array_get_idx (tracks, i);
+						json_object *trackInfo = NULL;
+						PianoSong_t *song;
+
+						if ((song = calloc (1, sizeof (*song))) == NULL) {
+							return PIANO_RET_OUT_OF_MEMORY;
+						}
+
+						song->seedId = PianoJsonStrdup(result, "pandoraId");
+						song->trackToken = PianoJsonStrdup (s, "trackPandoraId");
+						assert (song->trackToken != NULL);
+
+						LOG("track %d: %s\n",i + 1,song->trackToken);
+
+						if (!json_object_object_get_ex (annotations, song->trackToken, &trackInfo)) {
+							break;
+						}
+						assert (trackInfo!= NULL);
+						song->artist = PianoJsonStrdup(trackInfo, "artistName");
+						song->album = PianoJsonStrdup(trackInfo, "albumName");
+						song->title = PianoJsonStrdup(trackInfo, "name");
+						song->fileGain = 0.0;
+						song->length = getInt(trackInfo, "duration");
+						song->coverArt = getCoverArt(trackInfo);
+						playlist = PianoListAppendP (playlist, song);
+					}
+					break;
+				}
+
+				case PIANO_TYPE_ALBUM: {
+					char trackTitle[120];
+					int totalTracks = 0;
+					int trackNumber;
+
+				// Count tracks
+					json_object_object_foreach(result,Key1,Val1) {
+						if(Key1[0] != 'T' || Key1[1] != 'R') {
+							continue;
+						}
+						totalTracks++;
+					}
+
+					json_object_object_foreach(result,Key,Val) {
+						if(Key[0] != 'T' || Key[1] != 'R') {
+							continue;
+						}
+						LOG("got track %s: ",Key);
+						trackNumber = getInt(Val,"trackNumber");
```

Comment:
It this is the album track, we could probably extend `PianoSong_t` and add a `albumTrackNumber` field, so we don’t have to hard-code this here. When listing an album with `G`, it’ll be numbered anyways.